node/core/approval-voting/src/tests.rs

```diff
@@ -466,3 +466,797 @@ async fn overseer_signal(
+
+		virtual_overseer
+	});
+}
+
+#[test]
+fn ss_rejects_assignment_with_unknown_candidate() {
+	let (oracle, _handle) = make_sync_oracle(false);
+	test_harness(Default::default(), Box::new(oracle), |test_harness| async move {
+		let TestHarness {
+			mut virtual_overseer,
+			..
+		} = test_harness;
+
+		let block_hash = Hash::repeat_byte(0x01);
+		let candidate_index = 7;
+		let validator = ValidatorIndex(0);
+
+		// Add block hash 00.
+		import_n_blocks(&mut virtual_overseer, vec![block_hash]).await;
+
+		let rx = cai_assignment(
+			&mut virtual_overseer,
+			block_hash,
+			candidate_index,
+			validator,
+		).await;
+
+		assert_eq!(
+			rx.await,
+			Ok(AssignmentCheckResult::Bad(AssignmentCheckError::InvalidCandidateIndex(candidate_index))),
+		);
+
+		virtual_overseer
+	});
+}
+
+#[test]
+fn ss_accepts_and_imports_approval_after_assignment() {
+	let (oracle, _handle) = make_sync_oracle(false);
+	test_harness(Default::default(), Box::new(oracle), |test_harness| async move {
+		let TestHarness {
+			mut virtual_overseer,
+			..
+		} = test_harness;
+
+		let block_hash = Hash::repeat_byte(0x01);
+
+		let candidate_hash = {
+			let mut candidate_receipt = CandidateReceipt::<Hash>::default();
+			candidate_receipt.descriptor.para_id = 1.into();
+			candidate_receipt.descriptor.relay_parent = block_hash;
+			candidate_receipt.hash()
+		};
+
+		let candidate_index = 0;
+		let validator = ValidatorIndex(0);
+		let session_index = 1;
+
+		// Add block hash 0x01...
+		import_n_blocks(&mut virtual_overseer, vec![block_hash]).await;
+
+		let rx = cai_assignment(
+			&mut virtual_overseer,
+			block_hash,
+			candidate_index,
+			validator,
+		).await;
+
+		assert_eq!(rx.await, Ok(AssignmentCheckResult::Accepted));
+
+		let rx = cai_approval(
+			&mut virtual_overseer,
+			block_hash,
+			candidate_index,
+			validator,
+			candidate_hash,
+			session_index,
+		).await;
+
+		assert_eq!(rx.await, Ok(ApprovalCheckResult::Accepted));
+
+		virtual_overseer
+	});
+}
+#[test]
+fn ss_assignment_import_updates_candidate_entry_and_schedules_wakeup() {
+	let (oracle, _handle) = make_sync_oracle(false);
+	test_harness(Default::default(), Box::new(oracle), |test_harness| async move {
+		let TestHarness {
+			mut virtual_overseer,
+			..
+		} = test_harness;
+
+		let block_hash = Hash::repeat_byte(0x01);
+
+		let _candidate_hash = {
+			let mut candidate_receipt = CandidateReceipt::<Hash>::default();
+			candidate_receipt.descriptor.para_id = 1.into();
+			candidate_receipt.descriptor.relay_parent = block_hash;
+			candidate_receipt.hash()
+		};
+
+		let candidate_index = 0;
+		let validator = ValidatorIndex(0);
+
+		// Add block hash 0x01...
+		import_n_blocks(&mut virtual_overseer, vec![block_hash]).await;
+
+		let rx = cai_assignment(
+			&mut virtual_overseer,
+			block_hash,
+			candidate_index,
+			validator,
+		).await;
+
+		assert_eq!(rx.await, Ok(AssignmentCheckResult::Accepted));
+
+		// TODO(ladi): fix
```

Comment:
Having some trouble getting the wakeups scheduled. Needs a little more "thinking through" on my part.

Reply:
This needs an issue filed